File: EstructurasLineales/Java/Queue.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package ed.list;

/**
 *
 * @author A01411440
 * @param <T>
 */
public class Queue <T>{
    private Node<T> front;
    private Node<T> rear;
    private int size;
    
    Queue(){
        front = rear = null;
        size = 0;
    }
    Queue(T d){
        Node node = new Node (d);
        front = rear = node;
        size = 1;
    }
    public void enqueue(T d){
        Node node = new Node(d);
        if(isEmpty()){
            front=rear=node;
            size=1;
        }else{
            rear.next=node;
            rear = node;
            size++;
        }
    }
    public T dequeue(){
        if(isEmpty()){
            return null;
        }else{
            Node<T> fd = front;
            front= front.next;
            size--;
            return fd.data;
        }
    }
    public boolean isEmpty(){
        return size ==0;
    } 
    public T getFront(){
        if (isEmpty()){
        return null;
        }else{
            return front.data;
        }
    }
    public int getSize(){
        return size;
    }
    public void clearQueue(){
        front = rear = null;
        size = 0;
    }
    public void showQueue(){
        System.out.print("front->"); 
        Node<T> p =front;
        if(p==null){
            System.out.print("NULL");
        }
        while(p!=null){
            p.print();
            p=p.next;
        }
        System.out.println("<-rear");
    }
}

File: EstructurasLineales/C++/Queue.cpp
#include "Queue.h"

template <typename T>
Queue<T>::Queue() {
	front = rear = NULL;
	size = 0;
}

template <typename T>
Queue<T>::Queue(T d) {//Al ingresar instanciamos un nodo y hacemos que front y rear apunten al nuevo nodo.
	Node<T> *node = new Node<T>(d);
	front = rear = node;
	size = 1;
}

template <typename T>
Queue<T>::~Queue() {}

template <typename T>
void Queue<T>::enqueue(T d) {
	Node<T> *node = new Node<T>(d);
	if (isEmpty()) {//Si esta vacio
		front = rear = node;
		size = 1;
	}
	else {
		rear->next = node;
		rear = node;
		size++;
	}
}

template <typename T>
T Queue<T>::dequeue() {//QUITAR (ELIMINA Y DEVUELVE)
	if (isEmpty()) {
		return NULL;
	}
	else {
		Node<T> *fd = front; //Declara un apuntador
		front = front->next; //Avanza front
		size--;//Reduce el tamaño
		return fd->data; //Regresa el dato del nodo borrado
	}
}

template <typename T>
T Queue<T>::getFront() {
	if (isEmpty()) {
		return NULL;
	}
	else {
		return front->data;
	}
}

template <typename T>
int Queue<T>::getSize() {
	return size;
}

template <typename T>
bool Queue<T>::isEmpty() {
	return size == 0;
}

template <typename T>
void Queue<T>::clearQueue() {
	front = rear = NULL;
	size = 0;
}

template <typename T>
void Queue<T>::showQueue() {
	std::cout << "front->";
	Node<T>* p = front;
	if (!p) std::cout << "NULL"; //Si el puntero es nulo es FALSO (aqui se niega el nulo por lo cual es VERDADERO)
	while (p) {
		p->print();
		p = p->next;
	}
	std::cout << "<-rear" << endl;
}